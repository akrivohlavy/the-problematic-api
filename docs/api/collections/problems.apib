## Problems [/problems]

+ Attributes (array[Problem.Base])

### List All Problems [GET]

+ Parameters
    + type (string, optional) - `"riddle"` or `"expression"`
    + answered (boolean, optional)

+ Request
    + Headers
        Authorization: Basic ABC123

+ Response 200 (application/json)
    + Attributes (Problems)

