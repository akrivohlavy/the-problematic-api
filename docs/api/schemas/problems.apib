
## Problem (object)
+ id: 1 (number) - An unique identifier of the Problem
+ answered (boolean) - Does the question have any answers?


## Problem.Create (object)
+ type: `riddle` (string, required)

    Body of the problem

+ query: `If two is a company and three is a crowd. What is four and five?` (string, required)
