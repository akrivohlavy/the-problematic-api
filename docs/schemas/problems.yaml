type: array
items:
  $ref: './problem.yaml'
