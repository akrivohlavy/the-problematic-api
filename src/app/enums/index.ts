// business model enums
export enum ProblemType {
    riddle,
    expression,
}

export default {
    // general enums
};
