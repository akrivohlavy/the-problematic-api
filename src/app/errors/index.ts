import { mapValues } from 'lodash';

enum ErrorCodes {
    TEMPLATE_TEST = 'Unicorn cries',
    // general errors
    e0000 = "I'm a teapot",
    e0001 = 'Question already is the same.',
    e0002 = 'Unknown problem type.',

    e4000 = 'Validation error.',
    e4001 = 'Not Authenticated.',
    e4002 = 'Only user that created a Problem is allowed to modify it.',
}

// Type hacking for simplified error notation and code suggestions
export const E_CODES = (mapValues(
    ErrorCodes,
    (message: string, code: string): any => ({ message, code })
) as any) as typeof ErrorCodes;
