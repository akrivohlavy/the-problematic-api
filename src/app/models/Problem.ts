import { DataTypes, Model } from 'sequelize';
import { sequelize } from './index';

export default class Problem extends Model {
    public id!: number;
    public type!: string;
    public query!: string;
    public createdBy!: string;
    public answered!: boolean;
}

Problem.init(
    {
        id: {
            type: DataTypes.INTEGER,
            autoIncrement: true,
            primaryKey: true,
        },
        type: {
            type: DataTypes.STRING,
            allowNull: false,
        },
        query: {
            type: DataTypes.STRING,
            allowNull: false,
        },
        createdBy: {
            type: DataTypes.STRING,
            allowNull: false,
        },
        answered: {
            type: DataTypes.BOOLEAN,
        },
    },
    {
        sequelize,
        tableName: 'problems',
    }
);
