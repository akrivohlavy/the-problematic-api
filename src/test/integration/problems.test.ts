import * as request from 'supertest-as-promised';

import app from 'server';
import Problem from '../../app/models/Problem';
import { aBetterProblemMock, createProblemMock, expressionMock, otherUserMock, userMock } from './__mocks__/problems';

describe('Problems (Integration)', () => {
    describe('API', () => {
        it('is protected by Basic Auth', async () => {
            await request(app)
                .get('/api/problems')
                .send(createProblemMock)
                .expect(401)
                .then(({ headers }: any) => {
                    expect(headers).toHaveProperty('www-authenticate');
                    expect(headers['www-authenticate']).toStartWith('Basic');
                });

            await request(app)
                .get('/api/problems')
                .auth(userMock.username, userMock.password)
                .then(({ status }: any) => {
                    expect(status)
                        .toBeGreaterThanOrEqual(200)
                        .toBeLessThanOrEqual(299);
                });
        });
    });

    describe('User can', () => {
        let tempRiddle: Problem;
        let tempExpression: Problem;

        it('create a riddle Problem', () => {
            return request(app)
                .post('/api/problems')
                .auth(userMock.username, userMock.password)
                .send(createProblemMock)
                .expect(200)
                .then(({ body }: any) => {
                    expect(body).toMatchSnapshot();
                    tempRiddle = body;
                });
        });

        it('create an expression Problem', () => {
            return request(app)
                .post('/api/problems')
                .auth(userMock.username, userMock.password)
                .send(expressionMock)
                .expect(200)
                .then(({ body }: any) => {
                    expect(body).toMatchSnapshot();
                    tempExpression = body;
                });
        });

        it('list Problems', () => {
            return request(app)
                .get('/api/problems')
                .auth(userMock.username, userMock.password)
                .expect(200)
                .then(({ body }: any) => {
                    expect(body).toMatchSnapshot();
                });
        });

        it('list Problems filtered by type', () => {
            return request(app)
                .get('/api/problems')
                .query({ type: 'riddle' })
                .auth(userMock.username, userMock.password)
                .expect(200)
                .then(({ body }: any) => {
                    expect(body).toMatchSnapshot();
                });
        });

        it('read Problem', () => {
            return request(app)
                .get(`/api/problems/${tempRiddle.id}`)
                .auth(userMock.username, userMock.password)
                .expect(200)
                .then(({ body }: any) => {
                    expect(body).toEqual(tempRiddle);
                });
        });

        it('update Problem', async () => {
            await request(app)
                .put(`/api/problems/${tempRiddle.id}`)
                .auth(userMock.username, userMock.password)
                .send(aBetterProblemMock)
                .expect(200)
                .then(({ body }: any) => {
                    expect(body.query).toEqual(aBetterProblemMock.query);
                    expect(body.id).toEqual(tempRiddle.id);
                });

            await request(app)
                .put(`/api/problems/${tempRiddle.id}`)
                .auth(userMock.username, userMock.password)
                .send(aBetterProblemMock)
                .expect(304);
        });

        it('not modify Problem created by other user', async () => {
            await request(app)
                .put(`/api/problems/${tempRiddle.id}`)
                .auth(otherUserMock.username, otherUserMock.password)
                .send(createProblemMock)
                .expect(403);
        });

        it('not delete Problem created by other user', async () => {
            await request(app)
                .delete(`/api/problems/${tempRiddle.id}`)
                .auth(otherUserMock.username, otherUserMock.password)
                .send(createProblemMock)
                .expect(403);
        });

        it('delete Problems', async () => {
            await request(app)
                .delete(`/api/problems/${tempRiddle.id}`)
                .auth(userMock.username, userMock.password)
                .expect(204);

            await request(app)
                .delete(`/api/problems/${tempExpression.id}`)
                .auth(userMock.username, userMock.password)
                .expect(204);

            return await request(app)
                .get('/api/problems')
                .auth(userMock.username, userMock.password)
                .expect(204)
                .then(({ body }: any) => {
                    expect(body).toBeEmpty();
                });
        });
    });
});
