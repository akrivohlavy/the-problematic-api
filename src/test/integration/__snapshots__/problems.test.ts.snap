// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Problems (Integration) User can create a riddle Problem 1`] = `
Object {
  "answered": false,
  "createdBy": "user1",
  "id": 1,
  "query": "How much wood would a woodchuck chuck if a woodchuck could chuck wood?",
  "type": "riddle",
}
`;

exports[`Problems (Integration) User can create an expression Problem 1`] = `
Object {
  "answered": false,
  "createdBy": "user1",
  "id": 2,
  "query": "5+5",
  "type": "expression",
}
`;

exports[`Problems (Integration) User can list Problems 1`] = `
Array [
  Object {
    "answered": false,
    "createdBy": "user1",
    "id": 1,
    "query": "How much wood would a woodchuck chuck if a woodchuck could chuck wood?",
    "type": "riddle",
  },
  Object {
    "answered": false,
    "createdBy": "user1",
    "id": 2,
    "query": "5+5",
    "type": "expression",
  },
]
`;

exports[`Problems (Integration) User can list Problems filtered by type 1`] = `
Array [
  Object {
    "answered": false,
    "createdBy": "user1",
    "id": 1,
    "query": "How much wood would a woodchuck chuck if a woodchuck could chuck wood?",
    "type": "riddle",
  },
]
`;
